### PR TITLE
Let intersphinx links to Minimum/Maximum work by putting them in the docs

### PR DESCRIPTION
**Pull request by jamadden**
Also expose `AbstractValue` from zope.minmax; it was already documented as such.

I noticed these issues while documenting `zope.session`.

### CODE REVIEW

**Comment by zvezdan** [comment]
Please use a more standardized docstring format.

```
"""Summary sentence cuddled with triple quotes.

Long description paragraph after the blank line.

Empty line after the last paragraph so that Emacs/Vi users
can rewrap it with one keystroke.

"""
```

The use of `pkg_resources` in docs/conf.py feels like overkill too. I'll take a look over the weekend if there are better settings.
 
Finally, I need to go a decade back in time when I wrote this to see if `AbstractValue` was actually intended to be public.

**Comment by jamadden** [comment]
Thank you for your comments.

1. The added docstrings are a common variant style used by e.g., [django](https://docs.djangoproject.com/en/1.11/ref/contrib/admin/admindocs/#model-reference). That format is already [in use](https://github.com/zopefoundation/zope.traversing/blob/master/src/zope/traversing/interfaces.py#L104) in [a variety](https://github.com/zopefoundation/zope.intid/blob/master/src/zope/intid/interfaces.py#L6) of [zopefoundation projects](https://github.com/zopefoundation/zope.browserresource/blob/master/src/zope/browserresource/interfaces.py#L21). (Granted, I probably added a bunch of those, but I wasn't the first 😄 and reviewers typically haven't complained. )

    The use of an extra trailing newline does not seem to be widespread. As an Emacs user I can attest that it's not necessary there for wrapping to work with M-q. Perhaps @mgedmin can comment about vim (but that hasn't come up before either).

2. The use of `pkg_resources` is widespread in newer documentation projects to keep the version numbers in sync, as suggested by tseaver (off the top of my head, at least [zope.viewlet](https://github.com/zopefoundation/zope.viewlet/blob/master/docs/conf.py) [zope.browserresource](https://github.com/zopefoundation/zope.browserresource/blob/master/docs/conf.py), [zope.contentprovider](https://github.com/zopefoundation/zope.contentprovider/blob/master/docs/conf.py), [zope.testrunner](https://github.com/zopefoundation/zope.testrunner/blob/master/docs/conf.py) and many more do so --- of course, once again, many of those I added after that was pointed out to me in a review). The alternative is either complicated regexes or a `version.txt` file, which [was frowned upon](https://github.com/zopefoundation/zope.container/pull/8#discussion_r92414379)

3. `AbstractValue` is de-facto [documented as public](https://zopeminmax.readthedocs.io/en/latest/narrative.html). I know for sure that [third-party packages](http://ntizodb.readthedocs.io/en/latest/api.html#module-nti.zodb.minmax) are extending it.

**Comment by zvezdan** [comment]
Let me clarify:

1. If you feel strongly about using this docstring style, I'm not going to argue. All I'm saying is that in the original implementation I would have written it the way I requested above. Feel free to drop my suggestion.

2. I understand it's important to automate version fetching. It's probably OK to use `pkg_resources` for the documentation builds, since the baggage they bring is limited only to **build time**. In runtime, that baggage is much more impactful, so I tend to look for different options there. Again, go for it ...

3. I wrote the original documentation as a doctest file. That was later moved into a Sphinx doc and this reference exposed there. It does not mean it was intended to be public.

In fact, the imports in `__init__.py` for `Maximum` and `Minimum` indicate that they were the only publicly exposed classes (besides the interface module). The module was deliberately named `_minmax.py` (leading underscore) to indicate privacy. We do not use leading underscores on class names in Python as we do on attributes/bindings (variables) to hint they're private. So, this "private" module was the way to do that.

The fact that this 3rd party package is importing a class from private module and using it, also does not necessarily mean that's a correct or intended usage. Python does not enforce private/public qualifiers and we can do a lot of things that go **around** the intended usage pattern.

That said, the things are not set in stone. If the users find this abstract class useful, and it's used in more than this one 3rd party package, then perhaps it should be public. I'd be curious to see if this is a one-off exception of reaching into the private module, or the more widespread usage pattern. The latter would mean the consumers want the API change. Otherwise, we at least need to entertain the possibility that the original intent of keeping it private was correct.

**Comment by jamadden** [comment]
Thank you very much for your clarifications and comments. I'd like to respond with some of my own.

Some more points regarding the publicness of `AbstractValue`:

I'll re-iterate that it has been publicly documented on RTD since [2015](https://github.com/zopefoundation/zope.minmax/commit/e1397fcb1631815c68b4c31c41ad382e393303a4). Even before that, the reference in the documentation file referred to `zope.minmax.AbstractValue` (note the missing underscore). People do read .txt files that are part of distributions, and zope.app.apidoc [does publish them](https://github.com/zopefoundation/zope.app.apidoc/blob/master/src/zope/app/apidoc/codemodule/module.py#L139), so I would argue that it has been documented as public long before that. That original documentation has been carried over into places like [FreeBSD](https://github.com/freebsd/freebsd-ports-kde/blob/23d30b26b37b47a36da753a61ec9fcd2a52a5f30/devel/py-zope.minmax/pkg-descr).

The interface it implements (`zope.minmax.interfaces.IAbstractValue`) is public and *often* (though not always) that means that the default implementation should be as well.

I'd expect the developers of that package expected it was an oversight/accident that it wasn't available as documented and went ahead and used it, *especially* after noticing in the REPL that `zope.minmax.Minimum` and `Maximum` were really just aliases for `zope.minmax._minmax.Minimum` et al. (And yeah, I'm the one that originally wrote that package---although I don't remember my exact thoughts at the time. Man, I seem to be a trouble maker 😄 )

Regarding pkg_resources:

The major costs of `pkg_resources` are borne [on import](https://github.com/pypa/setuptools/blob/master/pkg_resources/__init__.py#L3162) as it builds its caches. (That can definitely be slow if sys.path is long and/or contains lots of packages. I've felt that pain!). But `zope` is [a namespace package declared using `pkg_resources`](https://github.com/zopefoundation/zope.minmax/blob/master/src/zope/__init__.py) so both at doc build time and at runtime that cost has already been paid.

**Comment by zvezdan** [comment]
If you feel this train had left the station long time ago, go ahead with the change to public. :-)
Especially, if it's useful in your own packages.
It does seem that we have to keep that "private" module name (`_minmax.py`) because changing it could break existing users. Unless you want to bump a major version too.

File: docs/conf.py [hunk 1]
```diff
@@ -266,4 +269,11 @@
 
 
 # Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'http://docs.python.org/': None}
+intersphinx_mapping = {
+    'http://docs.python.org/': None,
```
Comment (mgedmin) [inline docs/conf.py]: https

**Review by mgedmin — APPROVED** [review]
LGTM

**Comment by jamadden** [comment]
Thanks for the thoughtful reviews!

**Pull request merged**